libraries/concurrent-merkle-tree/src/cmt.rs

```diff
@@ -0,0 +1,447 @@
+use crate::{
+    error::CMTError,
+    state::{ChangeLog, Node, Path, EMPTY},
+    utils::{empty_node, empty_node_cached, fill_in_proof, hash_to_parent, recompute},
+};
+use bytemuck::{Pod, Zeroable};
+pub(crate) use log_compute;
+pub(crate) use solana_logging;
+
+#[cfg(feature = "sol-log")]
+use solana_program::{log::sol_log_compute_units, msg};
+
+/// Enforce constraints on max depth and buffer size
+#[inline(always)]
+fn check_bounds(max_depth: usize, max_buffer_size: usize) {
+    assert!(max_depth < 31);
+    // This will return true if MAX_BUFFER_SIZE is a power of 2 or if it is 0
+    assert!(max_buffer_size & (max_buffer_size - 1) == 0);
+}
+
+/// Conurrent Merkle Tree is a Merkle Tree that allows
+/// multiple tree operations targeted for the same tree root to succeed.
+///
+/// In a normal merkle tree, only the first tree operation will succeed because the
+/// following operations will have proofs for the unmodified tree state. ConcurrentMerkleTree avoids
+/// this by storing a buffer of modified nodes (`change_logs`) which allows it to implement fast-forwarding
+/// of concurrent merkle tree operations. 
+///
+/// As long as the concurrent merkle tree operations
+/// have proofs that are valid for a previous state of the tree that can be found in the stored
+/// buffer, that tree operation's proof can be fast-forwarded and the tree operation can be
+/// applied. 
+///
+/// There are two primitive operations for Concurrent Merkle Trees:
+/// [set_leaf](ConcurrentMerkleTree:set_leaf) and [append](ConcurrentMerkleTree::append). Setting a leaf value requires
+/// passing a proof to perform that tree operation, but appending does not require a proof.
+///
+/// An additional key property of ConcurrentMerkleTree is support for [append](ConcurrentMerkleTree::append) operations, 
+/// which do not require any proofs to be passed. This is accomplished by keeping track of the
+/// proof to the rightmost leaf in the tree (`rightmost_proof`).
+#[derive(Copy, Clone)]
+pub struct ConcurrentMerkleTree<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> {
```

Comment:
I'm not sure if you're planning on making some docs for how to use this.  Are people meant to store these on-chain?

Reply:
Yes